File: apps/android/app/src/main/java/com/dndsync/android/sync/SyncSession.kt
package com.dndsync.android.sync

import android.util.Log
import com.dndsync.android.BuildConfig

import com.dndsync.android.lan.LanUiState
import com.dndsync.android.pair.UnpairContext
import com.dndsync.proto.v1.CloudEnvelope
import com.dndsync.proto.v1.DndState
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Originate and apply DndState. Owns echo suppression and last-write-wins
 * across LAN and the wake-and-pull CloudEnvelope.
 */
class SyncSession(
    private val lan: SyncLan,
    private val cloud: SyncCloud,
    private val pair: SyncPairing,
    private val nowMs: () -> Long = { System.currentTimeMillis() },
    mainDispatcher: CoroutineDispatcher = Dispatchers.Main.immediate,
    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
) {
    var onApplyRemote: ((Boolean) -> Unit)? = null

    private val scope = CoroutineScope(SupervisorJob() + mainDispatcher)
    private val ioScope = CoroutineScope(SupervisorJob() + ioDispatcher)
    private val gate = LanSyncGate(nowMs)
    private val nearbyGranted = AtomicBoolean(false)
    private var originGeneration = 0
    private var currentOrigin: Origin? = null

    private class Origin(val generation: Int) {
        val cloudStarted = AtomicBoolean(false)
        var lanJob: Job? = null
        var cloudJob: Job? = null
    }

    val lanUi: StateFlow<LanUiState> = lan.ui

    init {
        wire()
    }

    fun onLocalFocusChange(on: Boolean) {
        gate.noteObserverEvent()
        gate.onLocalChange(on)
    }

    fun noteObserverEvent() {
        gate.noteObserverEvent()
    }

    fun onInboundState(state: DndState, viaLan: Boolean) {
        if (gate.onRemote(state)) {
            pair.persistLastSync(
                unixMs = state.unixMs,
                on = state.on,
                sender = state.sender,
                viaLan = viaLan,
            )
        }
    }

    fun onInboundUnpair(control: com.dndsync.proto.v1.PairControl) {
        pair.handleInboundUnpair(control)
    }

    fun onInboundEnvelope(envelope: CloudEnvelope) {
        val raw = envelope.ciphertext.toByteArray()
        val plaintext: ByteArray = if (pair.joined) {
            val opened = pair.open(raw)
            if (opened != null) {
                opened
            } else if (envelope.payloadKind == CloudEnvelopeCodec.PAYLOAD_PAIR_CONTROL) {
                return
            } else {
                pair.noteCloudError("decrypt failed")
                return
            }
        } else {
            raw
        }
        val control = PairControlFrames.decode(plaintext)
        if (envelope.payloadKind == CloudEnvelopeCodec.PAYLOAD_PAIR_CONTROL || control != null) {
            if (control != null) {
                onInboundUnpair(control)
            }
            return
        }
        val state = try {
            DndState.parseFrom(plaintext)
        } catch (_: Exception) {
            pair.noteCloudError("invalid inner DndState")
            return
        }
        if (state.version != Wire.PROTO_VERSION) {
            return
        }
        onInboundState(state, viaLan = false)
    }

    fun onInboundCommand(on: Boolean) {
        val state = DndState.newBuilder()
            .setVersion(Wire.PROTO_VERSION)
            .setOn(on)
            .setUnixMs(nowMs())
            .setSender(Wire.SENDER_MAC)
            .build()
        onInboundState(state, viaLan = false)
    }

    fun sendUnpair(context: UnpairContext) {
        if (context.pairId.isEmpty()) {
            return
        }
        val control = PairControlFrames.unpair(pairId = context.pairId)
        if (context.pairSecret.isEmpty() || context.forwarderUrl.isEmpty()) {
            if (nearbyGranted.get()) {
                runBlocking(ioDispatcher) { lan.deliverUnpair(control) }
            }
            return
        }
        // Finish the forwarder notify before returning. PairSession.unpair()
        // stops LAN as soon as this returns, and a missed push is the only
        // way an off-LAN Mac learns about unpair.
        postUnpairAndDelete(context, control)
        if (nearbyGranted.get()) {
            runBlocking(ioDispatcher) { lan.deliverUnpair(control) }
        }
    }

    fun setPairId(pairId: String) {
        lan.setPairId(pairId)
    }

    fun setNearbyGranted(granted: Boolean) {
        nearbyGranted.set(granted)
        syncTransport()
    }

    fun startLANIfJoined() {
        syncTransport()
    }

    fun stopLAN() {
        lan.stop()
        gate.reset()
    }

    private fun syncTransport() {
        if (pair.joined && nearbyGranted.get()) {
            lan.start()
        } else {
            lan.stop()
        }
    }

    private fun wire() {
        gate.onOriginate = { state ->
            if (pair.joined) {
                val previous = currentOrigin
                previous?.lanJob?.cancel()
                if (previous?.cloudStarted?.get() != true) {
                    previous?.cloudJob?.cancel()
                }
                val origin = Origin(++originGeneration)
                currentOrigin = origin
                origin.lanJob = scope.launch {
                    val nearby = nearbyGranted.get()
                    debug("flip on=${state.on} unix_ms=${state.unixMs} nearby=$nearby")
                    val acked = try {
                        if (nearby) lan.deliverState(state) else false
                    } catch (cancelled: CancellationException) {
                        throw cancelled
                    } catch (error: Exception) {
                        debug("lan deliver failed ${error.message}")
                        throw error
                    }
                    if (!isActive || origin.generation != originGeneration) {
                        debug("flip dropped stale unix_ms=${state.unixMs}")
                        return@launch
                    }
                    if (acked) {
                        debug("lan ack unix_ms=${state.unixMs}")
                        pair.persistLastSync(
                            unixMs = state.unixMs,
                            on = state.on,
                            sender = state.sender,
                            viaLan = true,
                        )
                        return@launch
                    }
                    debug("cloud fallback unix_ms=${state.unixMs}")
                    origin.cloudStarted.set(true)
                    origin.cloudJob = ioScope.launch {
                        postEnvelope(state)
                        if (origin.generation == originGeneration) {
                            pair.persistLastSync(
                                unixMs = state.unixMs,
                                on = state.on,
                                sender = state.sender,
                                viaLan = false,
                            )
                        }
                    }
                }
            }
        }
        gate.onApplyRemote = { on ->
            onApplyRemote?.invoke(on)
        }
        lan.inbound
            .onEach { state -> onInboundState(state, viaLan = true) }
            .launchIn(scope)
        lan.inboundUnpair
            .onEach { control -> onInboundUnpair(control) }
            .launchIn(scope)
    }

    private fun postEnvelope(state: DndState) {
        if (!pair.joined || pair.forwarderURL.isEmpty() || pair.pairSecret.isEmpty()) {
            debug("cloud post skipped joined=${pair.joined}")
            return
        }
        ioScope.launch {
            try {
                val ciphertext = pair.seal(state.toByteArray())
                val envelope = CloudEnvelopeCodec.make(
                    pair.pairId,
                    Wire.SENDER_ANDROID,
                    ciphertext,
                )
                cloud.postEnvelope(
                    pair.forwarderURL,
                    pair.pairId,
                    pair.pairSecret,
                    envelope.toByteArray(),
                )
                debug("cloud post ok unix_ms=${state.unixMs}")
                pair.noteCloudSuccess()
            } catch (error: Exception) {
                debug("cloud post failed ${error.message}")
                if (error is com.dndsync.android.cloud.ForwarderException.Unauthorized) {
                    pair.noteCloudUnauthorized()
                } else {
                    pair.noteCloudError(error.message ?: "request failed")
                }
            }
        }
    }

    private fun postUnpairAndDelete(
        context: UnpairContext,
        control: com.dndsync.proto.v1.PairControl,
    ) {
        try {
            val inner = control.toByteArray()
            val ciphertext = context.aesKey?.let {
                com.dndsync.android.pair.E2ECrypto.encrypt(inner, it)
            } ?: inner
            val envelope = CloudEnvelopeCodec.make(
                context.pairId,
                Wire.SENDER_ANDROID,
                ciphertext,
                CloudEnvelopeCodec.PAYLOAD_PAIR_CONTROL,
            )
            cloud.postEnvelope(
                context.forwarderUrl,
                context.pairId,
                context.pairSecret,
                envelope.toByteArray(),
            )
        } catch (_: Exception) {
        }
        try {
            cloud.deletePair(context.forwarderUrl, context.pairId, context.pairSecret)
        } catch (_: Exception) {
        }
    }

    private fun debug(message: String) {
        if (!BuildConfig.DEBUG) return
        try {
            Log.d("DndSync", "[DEBUG] $message")
        } catch (_: RuntimeException) {
            // JVM unit tests have no Android log.
        }
    }
}

File: apps/macos/DNDSync/SyncSession.swift
import Foundation

/// Originate and apply DndState. Owns echo suppression and last-write-wins
/// across LAN and the wake-and-pull CloudEnvelope.
@MainActor
final class SyncSession {
    var lanAdvertising = false
    var lanBrowsing = false
    var lanConnected = false
    var lastInboundText = "Last inbound: —"
    var lastLanErrorText = "Last LAN error: —"
    var lastEnvelopePostText = "Last envelope POST: —"

    var onStateChange: (() -> Void)?
    var onApplyRemote: ((Bool) -> Void)?

    private let lan: SyncLan
    private let cloud: SyncCloud
    private let pair: SyncPairing
    private let gate: LanSyncGate

    init(
        lan: SyncLan,
        cloud: SyncCloud,
        pair: SyncPairing,
        nowMs: @escaping () -> Int64 = {
            Int64(Date().timeIntervalSince1970 * 1000)
        }
    ) {
        self.lan = lan
        self.cloud = cloud
        self.pair = pair
        self.gate = LanSyncGate(sender: LanConstants.senderMac, nowMs: nowMs)
        wire()
    }

    func onLocalFocusChange(on: Bool) {
        gate.noteObserverEvent()
        gate.onLocalChange(on: on)
    }

    func onInboundState(_ state: Dndsync_V1_DndState, viaLan: Bool) {
        lastInboundText =
            "Last inbound: on=\(state.on) unix_ms=\(state.unixMs) sender=\(state.sender)"
        publish()
        if gate.onRemote(state) {
            debug("apply on=\(state.on) unix_ms=\(state.unixMs) viaLan=\(viaLan)")
            pair.persistLastSync(
                unixMs: state.unixMs,
                on: state.on,
                sender: state.sender,
                viaLan: viaLan
            )
        } else {
            debug("drop on=\(state.on) unix_ms=\(state.unixMs) viaLan=\(viaLan)")
        }
    }

    func onInboundUnpair(_ control: Dndsync_V1_PairControl) {
        pair.handleInboundUnpair(control)
    }

    func onInboundEnvelope(_ envelope: Dndsync_V1_CloudEnvelope) {
        debug("cloud envelope kind=\(envelope.payloadKind) sender=\(envelope.sender)")
        let plaintext: Data
        if pair.joined {
            if let opened = pair.open(envelope.ciphertext) {
                plaintext = opened
            } else if envelope.payloadKind == CloudEnvelopeCodec.payloadPairControl {
                debug("cloud drop pair-control decrypt")
                return
            } else {
                debug("cloud drop decrypt")
                pair.noteCloudError("decrypt failed")
                return
            }
        } else {
            plaintext = envelope.ciphertext
        }
        if envelope.payloadKind == CloudEnvelopeCodec.payloadPairControl
            || PairControlFrames.decode(plaintext) != nil
        {
            if let control = PairControlFrames.decode(plaintext) {
                onInboundUnpair(control)
            }
            return
        }
        guard let state = DndStateFrames.decode(plaintext) else {
            debug("cloud drop invalid DndState")
            pair.noteCloudError("invalid inner DndState")
            return
        }
        onInboundState(state, viaLan: false)
    }

    private func debug(_ message: String) {
        #if DEBUG
        print("[DEBUG] \(message)")
        #endif
    }

    func sendUnpair(_ context: UnpairContext) {
        guard !context.pairId.isEmpty else {
            return
        }
        let control = PairControlFrames.unpair(pairId: context.pairId)
        lan.sendUnpairNow(control)
        guard !context.pairSecret.isEmpty, !context.forwarderURL.isEmpty else {
            return
        }
        Task {
            await self.postUnpairAndDelete(context: context, control: control)
        }
    }

    func setPairId(_ pairId: String) {
        lan.setPairId(pairId)
    }

    func startLANIfJoined() {
        if pair.joined {
            lan.start()
        }
    }

    func stopLAN() {
        lan.stop()
        gate.reset()
    }

    private func wire() {
        gate.onOriginate = { [weak self] state in
            guard let self, self.pair.joined else { return }
            self.postEnvelope(state)
            self.pair.persistLastSync(
                unixMs: state.unixMs,
                on: state.on,
                sender: state.sender,
                viaLan: false
            )
        }
        gate.onApplyRemote = { [weak self] on in
            self?.onApplyRemote?(on)
        }
        lan.onUiState = { [weak self] snapshot in
            Task { @MainActor in
                guard let self else { return }
                self.lanAdvertising = snapshot.advertising
                self.lanBrowsing = snapshot.browsing
                self.lanConnected = snapshot.connected
                if let error = snapshot.lastError {
                    self.lastLanErrorText = "Last LAN error: \(error)"
                } else if snapshot.connected {
                    self.lastLanErrorText = "Last LAN error: —"
                }
                self.publish()
            }
        }
        lan.onInbound = { [weak self] state in
            Task { @MainActor in
                self?.onInboundState(state, viaLan: true)
            }
        }
        lan.onInboundUnpair = { [weak self] control in
            Task { @MainActor in
                self?.onInboundUnpair(control)
            }
        }
    }

    private func postEnvelope(_ state: Dndsync_V1_DndState) {
        Task {
            do {
                let inner = try state.serializedData()
                let ciphertext = try pair.seal(inner)
                let envelope = CloudEnvelopeCodec.make(
                    pairId: pair.pairId,
                    sender: LanConstants.senderMac,
                    ciphertext: ciphertext
                )
                let body = try CloudEnvelopeCodec.encode(envelope)
                try await cloud.postEnvelope(
                    baseURL: pair.forwarderURL,
                    pairId: pair.pairId,
                    secret: pair.pairSecret,
                    envelope: body
                )
                lastEnvelopePostText = "Last envelope POST: 204"
                pair.noteCloudSuccess()
                publish()
            } catch {
                if isUnauthorized(error) {
                    pair.noteCloudUnauthorized()
                } else {
                    pair.noteCloudError(error.localizedDescription)
                }
                publish()
            }
        }
    }

    private func postUnpairAndDelete(context: UnpairContext, control: Dndsync_V1_PairControl) async {
        do {
            let inner = try control.serializedData()
            let ciphertext: Data
            if let aesKey = context.aesKey {
                ciphertext = try E2ECrypto.encrypt(plaintext: inner, key: aesKey)
            } else {
                ciphertext = inner
            }
            let envelope = CloudEnvelopeCodec.make(
                pairId: context.pairId,
                sender: LanConstants.senderMac,
                ciphertext: ciphertext,
                payloadKind: CloudEnvelopeCodec.payloadPairControl
            )
            let body = try CloudEnvelopeCodec.encode(envelope)
            try await cloud.postEnvelope(
                baseURL: context.forwarderURL,
                pairId: context.pairId,
                secret: context.pairSecret,
                envelope: body
            )
        } catch {
        }
        do {
            try await cloud.deletePair(
                baseURL: context.forwarderURL,
                pairId: context.pairId,
                secret: context.pairSecret
            )
        } catch {
        }
    }

    private func isUnauthorized(_ error: Error) -> Bool {
        if case ForwarderClientError.httpStatus(401, _) = error {
            return true
        }
        return error.localizedDescription.contains("HTTP 401")
    }

    private func publish() {
        onStateChange?()
    }
}
